### ISSUE TITLE
OData actions with parameters not registering without ODataActionParameters as another parameter on the controller action

### ISSUE THREAD

**Issue by joshcomley** [OData/AspNetCoreOData#426]
Hi there,

Using:
.NET 6
OData 8.0.5

Unless I am missing something, I am unable to get actions or functions with parameters to register properly unless I add an extra parameter to the controller action for `ODataActionParameters`.

To repro, clone the repo below and run:
[https://github.com/joshcomley/ODataActionBug](https://github.com/joshcomley/ODataActionBug)

For reference here is the `EdmModel` code:

    public static IEdmModel GetEdmModel()
    {
        if (_model != null)
        {
            return _model;
        }

        ODataConventionModelBuilder builder = new();
        
        var users = builder.EntitySet<User>("Users");

        // Works
        var hi = users.EntityType.Collection.Action(nameof(UsersController.Hi));
        hi.Returns<Hello>();

        // Doesn't work
        var hiSimpleFunction = users.EntityType.Collection.Function(nameof(UsersController.HiSimpleFunction));
        hiSimpleFunction.Returns<Hello>();
        hiSimpleFunction.Parameter<string>("name");

        // Doesn't work
        var hiSimple = users.EntityType.Collection.Action(nameof(UsersController.HiSimple));
        hiSimple.Returns<Hello>();
        hiSimple.Parameter<string>("name");

        // Doesn't work
        var hello = users.EntityType.Collection.Action(nameof(UsersController.Hello));
        hello.Returns<Hello>();
        hello.Parameter<SayHello>("request");

        // Doesn't work
        var goodbye = users.EntityType.Action(nameof(UsersController.Hello));
        goodbye.Returns<Hello>();
        goodbye.Parameter<SayHello>("request");

        return _model = builder.GetEdmModel();
    }

And one of the actions that isn't working, for example:

    [HttpPost]
    public IActionResult HiSimpleFunction(string name)
    {
        return Ok(
            new Hello
            {
                Message = $"Hi {name}!"
            }
        );
    }

**Comment by xuzhg** [comment]
@joshcomley 

1) Function should use "HttpGet' method. Change [HttpPost] to [HttpGet] on HiSimpleFunction should make it working.

2) For actions, you should add 'ODataActionParameters'. I am sorry the [FromODataBody] is not finished yet and hasn't been considered into routing builder yet.

**Comment by joshcomley** [comment]
@xuzhg thanks for your reply

1. There are other examples in the repo, surely actions and functions can be `HttpPost`?
2. So it is necessary to add `ODataActionParameters` on any action? Even if this parameter is unused? This is problematic as automatic code-cleanup can remove such unused parameters. An attribute of `[ODataAction]` would be better on the method.

**Comment by vasiliy-fofanov-stenn** [comment]
@xuzhg, when you're planning to implement `[FromODataBody]`?

It will be really helpful for Swagger generation because now with `ODataActionParameters` I can get only
next result

![image](https://user-images.githubusercontent.com/89073829/149140049-2d8b6dd3-7353-40ea-acd7-aa8e242fb896.png)

BTW, full example with Swagger Generation you can find [here](https://github.com/StennGroup/stenn-aspnetcore-versioning/tree/master/samples)

**Comment by vasiliy-fofanov-stenn** [comment]
> @xuzhg thanks for your reply
> 
> 1. There are other examples in the repo, surely actions and functions can be `HttpPost`?
> 2. So it is necessary to add `ODataActionParameters` on any action? Even if this parameter is unused? This is problematic as automatic code-cleanup can remove such unused parameters. An attribute of `[ODataAction]` would be better on the method.

1. Functions can be only `[HttGet]` and Actions - only `[HttpPost]`
2. Actions can be without parameter or contains only one parameter with type ODataActionParameters. Look at line 66 [here](https://github.com/OData/AspNetCoreOData/blob/69eec03c7003fe12d92cdc619efdc16781683694/src/Microsoft.AspNetCore.OData/Routing/Conventions/ActionRoutingConvention.cs)

**Comment by joshcomley** [comment]
Interesting. For me, as per the repo, `[FromODataBody]` seems to be working on actions, but my signature has to have both the model parameter *and* the `ODataActionParameters` parameter for it to be detected:

    [HttpPost]
    public async Task<string> Hello(
        [FromODataBody] HelloModel model, // This comes through as expected
        ODataActionParameters parameters // But nothing works if I don't ALSO add this parameter
    )
    {
        return $"Hello {model.Name}";
    }

**Comment by joshcomley** [comment]
Ah @vasiliy-fofanov-stenn yes I can see the requirement for the parameter as you pointed out. How odd that for me, the body is working fine

**Comment by joshcomley** [comment]
So to be clear...

This works:

    [HttpPost]
    public string Hello(
        [FromODataBody] HelloModel model,
        ODataActionParameters parameters
    )
    {
        return $"Hello {model.Name}";
    }
    
These don't work:

    [HttpPost]
    public string Hello(
        [FromBody] HelloModel model,
        ODataActionParameters parameters
    )
    {
        return $"Hello {model.Name}";
    }
    
    [HttpPost]
    public string Hello(
        HelloModel model,
        ODataActionParameters parameters
    )
    {
        return $"Hello {model.Name}";
    }
    
    [HttpPost]
    public string Hello(
        [FromODataBody] HelloModel model
    )
    {
        return $"Hello {model.Name}";
    }
    
    [HttpPost]
    public string Hello(
        [FromBody] HelloModel model
    )
    {
        return $"Hello {model.Name}";
    }
    
    [HttpPost]
    public string Hello(
        HelloModel model
    )
    {
        return $"Hello {model.Name}";
    }
    
In other words, `[FromODataBody]` is actually working and making a difference 🤷‍♂️

**Comment by vasiliy-fofanov-stenn** [comment]
> [FromODataBody]

@joshcomley, Are you sure that you call OData api method? 
Do you added action's parameter to edm model?

I tried your sample and it doesn't work on my side(( I got 500 when I try to call endpoint

Anyway, better to use [documented behavior](https://docs.microsoft.com/en-us/odata/webapi/action-parameter-support) for avoid problems during future updates

**Comment by sherlock1982** [comment]
This would be a fantastic feature so we can properly add validations there
